components/eamxx/src/physics/cosp/cosp_functions.hpp

```diff
@@ -0,0 +1,96 @@
+#ifndef SCREAM_COSP_FUNCTIONS_HPP
+#define SCREAM_COSP_FUNCTIONS_HPP
+#include "share/scream_types.hpp"
+using scream::Real;
+extern "C" void cosp_c2f_init(int ncol, int nsubcol, int nlay);
+extern "C" void cosp_c2f_final();
+extern "C" void cosp_c2f_run(int ncol, int nsubcol, int nlay, int ntau, int nctp,
+    Real emsfc_lw, Real* sunlit, Real* skt,
+    Real* T_mid, Real* p_mid, Real* p_int, Real* qv,
+    Real* cldfrac, Real* reff_qc, Real* reff_qi, Real* dtau067, Real* dtau105,
+    Real* isccp_cldtot, Real* isccp_ctptau);
+
+namespace scream {
+
+    namespace CospFunc {
+        using lview_host_1d = typename ekat::KokkosTypes<HostDevice>::template lview<Real*  >;
+        using lview_host_2d = typename ekat::KokkosTypes<HostDevice>::template lview<Real** >;
+        using lview_host_3d = typename ekat::KokkosTypes<HostDevice>::template lview<Real***>;
+        template <typename S>
+        using view_1d = typename ekat::KokkosTypes<HostDevice>::template view_1d<S>;
+        template <typename S>
+        using view_2d = typename ekat::KokkosTypes<HostDevice>::template view_2d<S>;
+        template <typename S>
+        using view_3d = typename ekat::KokkosTypes<HostDevice>::template view_3d<S>;
+
+        void initialize(int ncol, int nsubcol, int nlay) {
+            cosp_c2f_init(ncol, nsubcol, nlay);
+        };
+        void finalize() {
+            cosp_c2f_final();
+        };
+        void main(
+                Int ncol, Int nsubcol, Int nlay, Int ntau, Int nctp, Real emsfc_lw,
+                view_1d<const Real>& sunlit , view_1d<const Real>& skt,
+                view_2d<const Real>& T_mid  , view_2d<const Real>& p_mid  , view_2d<const Real>& p_int,
+                view_2d<const Real>& qv     , view_2d<const Real>& cldfrac,
+                view_2d<const Real>& reff_qc, view_2d<const Real>& reff_qi,
+                view_2d<const Real>& dtau067, view_2d<const Real>& dtau105,
+                view_1d<Real>& isccp_cldtot , view_3d<Real>& isccp_ctptau) {
+
+            // Make host copies and permute data as needed
+            lview_host_2d
+                  T_mid_h("T_mid_h", ncol, nlay), p_mid_h("p_mid_h", ncol, nlay), p_int_h("p_int_h", ncol, nlay+1),
+                  qv_h("qv_h", ncol, nlay), cldfrac_h("cldfrac_h", ncol, nlay),
+                  reff_qc_h("reff_qc_h", ncol, nlay), reff_qi_h("reff_qi_h", ncol, nlay),
+                  dtau067_h("dtau_067_h", ncol, nlay), dtau105_h("dtau105_h", ncol, nlay);
+            lview_host_3d isccp_ctptau_h("isccp_ctptau_h", ncol, ntau, nctp);
+            // NOTE: these should already be host views, so we could probably
+            // skip creating the mirror views here
+            auto sunlit_h = Kokkos::create_mirror_view(sunlit);
```

Comment:
Your `view_1d`/`view_2d` types are already on host. Why do you need to create mirrors for sunlit, skt, and isccp_cldtot?

Reply:
Carryover from when I was grabbing device views from the FM before you told me to just grab host views. Will fix.

Reply:
Fixed.